Update to support Kubernetes v1.6 (only) and add kubeadm support

luxas opened the pull request:
... for example on kubeadm clusters

fixes: #453 

Which versions of Kubernetes is Rook aiming at supporting?
I'd prefer if we could take the "latest-only" route initially while being in alpha, then we can move forward so much faster

Also, should we make Rook storageclass the default one by default? That would probably make sense, I think the case is more often that one does not have any other storageclasses when using rook compared to those who use rook and already have lots of other options.

In that case, we could just have a slightly different cmd for those who don't want rook to be the default...

kokhang commented on demo/kubernetes/rook-pvcontroller.yaml:
```diff
@@ -0,0 +1,48 @@
+  namespace: kube-system
+---
+apiVersion: extensions/v1beta1
+kind: Deployment
```
We might not have to replace the controller-manager. I know this is currently needed because Rook leverage the `rbd` plugin, which depends on ceph-tools. But im working on a rook volume plugin that will remove this dependency as well as adding some cool things like auto-mon discovery, ceph secret management and support for multiple mapping libraries (sysfc, ndb, iscsi)

kokhang commented on demo/kubernetes/rook-operator.yaml:
```diff
@@ -1,3 +1,59 @@
+# TODO: Update to rbac.authorization.k8s.io/v1beta1 once v1.5 support is dropped
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1alpha1
```
Thanks for doing this. We have been testing using the default service account and thus we have been able to access all these resources.

bassam commented:
> Which versions of Kubernetes is Rook aiming at supporting? I'd prefer if we could take the "latest-only" route initially while being in alpha, then we can move forward so much faster

@luxas I'd love to understand how long the tail is on K8S deployments. Is there any data on that anywhere? Would be great if we can say rook is for 1.7+

luxas commented:
@bassam as discussed at KubeCon, I hope that if rook can say it supports the most recent v1.x version while in alpha (for a given rook v0.x release), I think everyone is okay with that.

In that case, we could say that Rook v0.4.0 (whatever is coming soon) supports k8s v1.6 only, and I can use the latest stable API versions for everything, the right controller-manager binary, etc.

I can rebase this for v1.6 if you want pretty soon

bassam commented:
@luxas I agree. Rook should support latest K8S will in alpha -- so that means 1.6 for now.

luxas commented:
@bassam PTAL and test

kokhang commented on demo/kubernetes/rook-operator.yaml:
```diff
@@ -1,4 +1,61 @@
+kind: ServiceAccount
+metadata:
+  name: rook-operator
+  namespace: default
```
Any reason you are specifying the `default` namespace? This will not work if you create the rook-operator in a different namespace

luxas replied:
That's true. But in that case you have to change this yaml file anyway.
We have to choose one namespace for the operator and I think `default` is the best.

luxas replied:
Also it shouldn't matter where the rook operator is deployed. That's an issue with the operator, not the manifests.

kokhang replied:
Ok.. Sounds good. Do you think this would be clear for users or should we state in the docs that this yaml would only works under `default` namespace and if you want to deploy in a different namespace, you would change this yaml?

kokhang commented on demo/kubernetes/rook-pvcontroller.yaml:
```diff
@@ -0,0 +1,47 @@
+apiVersion: v1
```
This file would help us get over the RBD tools dependency. Once we get our external volume plugin, i hope this is not needed.

luxas replied:
This will not be needed once you build the Rook PV provisioner into the operator.
No need to build something into k8s for removing this.

luxas commented:
Updated now with more docs on this matter

kokhang commented on Documentation/k8s-pre-reqs.md:
```diff
@@ -1,6 +1,33 @@
-## New Local Kubernetes Cluster
+**WARNING:** Only Kubernetes v1.6.0 or higher is supported by Rook at this time (while Rook is in alpha it will track the latest release to use the latest features)
+
+## `kubeadm`
```
Any reason you are adding quotes to the header title? it just look different than the rest of the headers. I would use "## Kubeadm" just to keep it consistent.

kokhang commented on Documentation/k8s-pre-reqs.md:
```diff
@@ -9,19 +36,21 @@ export KUBECONFIG="$(pwd)/kubeconfig"
-Once you see a url response, your cluster is [ready for use by Rook](kubernetes.md#deploy-rook).
+Once you see an URL response, your cluster is [ready for use by Rook](kubernetes.md#deploy-rook).
+
+## `minikube`
```
Also here. Any reason to change the style of the header title?

luxas commented:
@kokhang @bassam Ready to merge!

dab-q commented on Documentation/k8s-pre-reqs.md:
````diff
@@ -9,19 +36,21 @@ export KUBECONFIG="$(pwd)/kubeconfig"
 ```
 
-Once you see a url response, your cluster is [ready for use by Rook](kubernetes.md#deploy-rook).
+Once you see an URL response, your cluster is [ready for use by Rook](kubernetes.md#deploy-rook).
````
grammar nit: It should be "a URL" (based on the pronunciation of the individual letters "you are ell", which doesn't start with a vowel sound, so it is "a", not "an")

jbw976 commented:
@luxas, I'm working on getting both our user guides and development workflow upgraded so that we support Kubernetes 1.6+.  We will probably wait to merge this PR until after that, since the changes in this PR require 1.6.  I will get this together ASAP.

jbw976 commented:
@luxas, I have updated the rook fork of coreos-kubernetes to use 1.6.1 so this PR should be able to proceed: https://github.com/rook/coreos-kubernetes/commit/a3e880bf3c880e4e4551fbba54d21cf6833ffb19

Please rebase your changes and I'll be happy to take them for a sanity test run :)

luxas commented on demo/kubernetes/rook-pvcontroller.yaml:
```diff
@@ -0,0 +1,47 @@
+      serviceAccountName: persistent-volume-binder
+      containers:
+      - name: pv-controller-manager
+        image: luxas/storage-controller-manager:v1.6.1
```
yes and the source is in this repo: https://github.com/luxas/kubeadm-workshop

luxas commented:
Rebased, lplease fix the grammar nit in a follow up

luxas commented:
No it's very late for me here (GMT+2), I just did a mechanical rebase as requested.

You can fix those in a follow up PR of yours, I don't think they are critical for v0.4

kokhang commented:
Hi @luxas, https://github.com/rook/rook/pull/577 was just merged. This added a volume provisioner to the operator. This means that we no longer need to patch the kube-controller to add `ceph-tools`. Do you think we could exclude the `rook-pvcontroller.yaml` from this PR and update the corresponding docs to reflect this? Thank you.

luxas commented:
I updated the PR to use the built-in provisioner (more RBAC rules, removed the other PV controller, etc.) and tested it end-to-end. Worked fine!
Also fixed the grammar nits so it should be good to go!

@bassam @kokhang @travisn

kwiesmueller commented:
I am getting the same error on a fresh 4 node k8s cluster on coreos (1.6.1) installed with kargo.
I did set the required RKT_OPTS and install RBD but the issue still appears.

```
Error from server (Forbidden): error when creating "rook-operator.yaml": clusterroles.rbac.authorization.k8s.io "rook-operator" is forbidden: attempt to grant extra privileges: [{[get] [] [namespaces] [] []} {[list] [] [namespaces] [] []} {[watch] [] [namespaces] [] []} {[create] [] [namespaces] [] []} {[update] [] [namespaces] [] []} {[get] [] [secrets] [] []} {[list] [] [secrets] [] []} {[watch] [] [secrets] [] []} {[create] [] [secrets] [] []} {[update] [] [secrets] [] []} {[get] [] [pods] [] []} {[list] [] [pods] [] []} {[watch] [] [pods] [] []} {[create] [] [pods] [] []} {[update] [] [pods] [] []} {[get] [] [services] [] []} {[list] [] [services] [] []} {[watch] [] [services] [] []} {[create] [] [services] [] []} {[update] [] [services] [] []} {[get] [] [nodes] [] []} {[list] [] [nodes] [] []} {[watch] [] [nodes] [] []} {[create] [] [nodes] [] []} {[update] [] [nodes] [] []} {[get] [] [configmaps] [] []} {[list] [] [configmaps] [] []} {[watch] [] [configmaps] [] []} {[create] [] [configmaps] [] []} {[update] [] [configmaps] [] []} {[get] [] [events] [] []} {[list] [] [events] [] []} {[watch] [] [events] [] []} {[create] [] [events] [] []} {[update] [] [events] [] []} {[get] [] [persistentvolumes] [] []} {[list] [] [persistentvolumes] [] []} {[watch] [] [persistentvolumes] [] []} {[create] [] [persistentvolumes] [] []} {[update] [] [persistentvolumes] [] []} {[get] [] [persistentvolumeclaims] [] []} {[list] [] [persistentvolumeclaims] [] []} {[watch] [] [persistentvolumeclaims] [] []} {[create] [] [persistentvolumeclaims] [] []} {[update] [] [persistentvolumeclaims] [] []} {[get] [extensions] [thirdpartyresources] [] []} {[list] [extensions] [thirdpartyresources] [] []} {[watch] [extensions] [thirdpartyresources] [] []} {[create] [extensions] [thirdpartyresources] [] []} {[get] [extensions] [deployments] [] []} {[list] [extensions] [deployments] [] []} {[watch] [extensions] [deployments] [] []} {[create] [extensions] [deployments] [] []} {[get] [extensions] [daemonsets] [] []} {[list] [extensions] [daemonsets] [] []} {[watch] [extensions] [daemonsets] [] []} {[create] [extensions] [daemonsets] [] []} {[get] [extensions] [replicasets] [] []} {[list] [extensions] [replicasets] [] []} {[watch] [extensions] [replicasets] [] []} {[create] [extensions] [replicasets] [] []} {[get] [storage.k8s.io] [storageclasses] [] []} {[list] [storage.k8s.io] [storageclasses] [] []} {[watch] [storage.k8s.io] [storageclasses] [] []} {[*] [rook.io] [*] [] []}] user=&{kube-admin-node1  [system:authenticated] map[]} ownerrules=[{[create] [authorization.k8s.io] [selfsubjectaccessreviews] [] []} {[get] [] [] [] [/api /api/* /apis /apis/* /healthz /swaggerapi /swaggerapi/* /version]}] ruleResolutionErrors=[]
```

kwiesmueller commented:
Hey,
i haven't explicitly set this. I might have to set it in my cert generation process, do you know if I have to do it there and replace the certs or is there another way of doing this?
